monai/transforms/compose.py

```diff
@@ -178,15 +178,45 @@ def randomize(self):
                     'Transform "{0}" in Compose not randomized\n{0}.{1}.'.format(type(_transform).__name__, type_error),
                     RuntimeWarning)
 
-    def __call__(self, input_):
-        for transform in self.transforms:
```

Comment:
The main goal of this PR is to propose a CacheDataset which will analyse a sequence of transforms in `Compose`, and cache the outcome of the deterministic transforms before the first non-deterministic one.

this specific code block tries to separate deterministic and non-deterministic transform calls in  `Compose`.

I feel this affects the code readability as the `__call__` method has quite complicated logic.  we are thinking about alternatives:

- option 1: adding `deterministic_call(data)` and `non_deterministic_call(data)` to  the `Compose` class.  which will return partially transformed data, so that the cachedataset can cache the deterministic ones.

- option 2: creating a `CacheCompose` class extending `Compose`, but this change requires that the user is familiar with `CacheCompose` and `CacheDataset`. with this, we couldn't achieve fancy things such as changing one line of user code, speed up training

this could be a major design change, so Nic and I would like to hold this PR and hear your opinion @ericspod

Reply:
Thanks @wyli and @ericspod for your comments.
I updated the code based on option 1, seems it's more clear and easy to use.
And added cache mechanism to integration test with determinism.
Could you please help review it again?
Thanks.